### PR TITLE
refactor: add postgres tailnet coordinator

### PR DESCRIPTION
**Pull request by spikecurtis**
c.f. #7295  and [RFC](https://www.notion.so/coderhq/Distributed-HA-Tailnet-Coordinator-cbc861d7d6494d079da88beff3096b86)

This PR adds a postgres-backed, enterprise, distributed tailnet coordinator.

Apologies for the large-ish PR, but this felt like the smallest amount of code that could be coherently understood and reviewed.

This PR adds the new coordinator, but leaves the old, pubsub-only coordinator, and does not wire up the new coordinator to anything.  Once this is merged, in a follow-on PR, I plan to introduce the postgres coordinator behind an experiment flag.  That will allow me to test the performance impact of the new coordinator in a scale test. Then, I suggest we enable the experiment on our dogfood instance and test for a while.

Other items missing from this PR that will be added later:

 * the ServeHTTPDebug endpoint
 * cleaning out orphaned state in the database from coordinators that crashed/exited before deleting their bindings

### CODE REVIEW

File: enterprise/tailnet/pgcoord.go
Comment (johnstcn) [inline enterprise/tailnet/pgcoord.go:653]: The locking logic here between `newConn` and `waitForConn` is tricky to follow. Is there a particular reason we lock, spawn a new goroutine, and unlock, only to immediately lock and unlock again?

Comment (spikecurtis) [reply enterprise/tailnet/pgcoord.go:653]: maybe a rename would make this more clear --- waitForConn doesn't do anything immediately, it waits for the connIO's context to expire, and then cleans it up.  It's on a separate goroutine so that we don't block newConn from returning.

How do you feed about `cleanupConn()`?

File: enterprise/tailnet/pgcoord.go
Comment (spikecurtis) [reply enterprise/tailnet/pgcoord.go:949]: the fact that its tied to `mKey | bKey` makes it not really reusable beyond this package.  Annoyingly, Go doesn't have a concept of "hashable" types, so it's hard to write this more generically than `mKey | bKey`.

Comment (johnstcn) [reply enterprise/tailnet/pgcoord.go:949]: :'(

File: enterprise/tailnet/pgcoord.go
Comment (johnstcn) [reply enterprise/tailnet/pgcoord.go:653]: Ah, the key here is `<-c.ctx.Done()` -- that's easy to skip over.
`cleanupConn` does make more sense in this context.

File: coderd/database/migrations/000127_ha_coordinator.up.sql
Comment (kylecarbs) [inline coderd/database/migrations/000127_ha_coordinator.up.sql:6]: Could this be integrated into the replicas table somehow? Feels unnecessary to have both essentially doing the same thing (heartbeating).

**Review by kylecarbs — APPROVED** [review]
I haven't had an opportunity to look through `pgcoord` thoroughly, but this looks good from a brief glance. My only comment is related to duplicated heartbeats!

File: enterprise/tailnet/pgcoord_test.go
Comment (coadler) [inline enterprise/tailnet/pgcoord_test.go:34]: Maybe it's just personal preference, but is there a reason you don't do nested tests?

File: enterprise/tailnet/pgcoord.go
Comment (coadler) [inline enterprise/tailnet/pgcoord.go:703]: `:many` queries like this won't ever return `sql.ErrNoRows`, only `:one`.

File: enterprise/tailnet/pgcoord_test.go
Comment (spikecurtis) [reply enterprise/tailnet/pgcoord_test.go:34]: Nesting increases the indent, and therefore leaves less horizontal room for readability, and https://github.com/coder/coder/pull/5376

File: enterprise/tailnet/pgcoord.go
Comment (mtojek) [inline enterprise/tailnet/pgcoord.go:791]: If `listenClient` expects only `pubsub.ErrDroppedMessages`, shouldn't it be extracted to another wrapping function?

File: enterprise/tailnet/pgcoord.go
Comment (mtojek) [inline enterprise/tailnet/pgcoord.go:764]: Should we add a sanity check to verify that this is a context related error? Otherwise we can log `BUG: ...`

File: enterprise/tailnet/pgcoord_test.go
Comment (mtojek) [inline enterprise/tailnet/pgcoord_test.go:114]: nit: actually it could be extracted to a helper function if we decide to replace `DB` with something else.

**Review by mtojek — APPROVED** [review]
_Actually, it is a new feature/implementation of the coordinator rather than refactoring. Keep in mind, that it is hard to give sensible feedback due to the size and complexity of the coordinator, but I did my best._

I went through the RFC first to understand implemented ideas like withdrawals or mappings. I have a few high-level questions:

1. Which database issues will be harmful to the PG coordinator? Are there ones he won't be able to recover without a restart?
2. Considering the number of bugs you're fixing, I have a feeling that we should open-source it instead of putting it in the `enterprise`. I assume that it is better than the in-memory implementation, right?
3. Did you think about the load testing plan to make sure that it works as intended and can handle expected traffic?
4. Security related: is it possible to sabotage the coordinator by sending fake/invalid queries? I know that this implementation is not purely distributed, so it can't be susceptible to typical Sybil attacks, but could it be possible to disconnect/isolate an agent or force it to connect to a node managed/taken over by an attacker?

**Comment by spikecurtis** [comment]
> 1. Which database issues will be harmful to the PG coordinator? Are there ones he won't be able to recover without a restart?

The usual suspects: data corruption, incompatible schema changes, etc.

In particular, it is _designed to recover gracefully_ from temporary loss of connectivity to the database.  When the pubsub reconnects, it triggers a resync of mappings.  Bindings errors are retried with a backoff.

**Comment by spikecurtis** [comment]
> Considering the number of bugs you're fixing, I have a feeling that we should open-source it instead of putting it in the enterprise. I assume that it is better than the in-memory implementation, right?

I think our current position is that features supporting multiple Coderd replicas are Enterprise licensed.

If you have only one Coderd, then the in-memory coordinator is totally sufficient, and will in fact have less latency to set up a new connection.

**Comment by spikecurtis** [comment]
> Did you think about the load testing plan to make sure that it works as intended and can handle expected traffic?

After merging, I plan to enable it behind an experiment flag and then run some load tests comparing it to the old `haCoordinator`.  I expect it to be a little more chatty with the database, on account of us pushing the state there, but hopefully not too different.  The goal here was correctness, not increased performance.

**Comment by spikecurtis** [comment]
> Security related: is it possible to sabotage the coordinator by sending fake/invalid queries? I know that this implementation is not purely distributed, so it can't be susceptible to typical Sybil attacks, but could it be possible to disconnect/isolate an agent or force it to connect to a node managed/taken over by an attacker?

The coordinator facilitates connectivity between clients and agents.  The Coderd API authenticates and authorizes connections into the coordinator, so by design we trust the information being provided by clients and agents. It's of course true that an employee or contractor could attempt to act maliciously, as an insider attack.  What could such a malicious actor do?

 1. Spam node updates to try to DoS the coordinator or database: They could certainly eat CPU on the coordinator server with us deserializing node updates over and over, but we only allow one DB upsert in flight for each connection, so we would limit the pressure on the database. 
 
 2. Send malformed data: This will result in deserialization errors, which triggers the coordinator to disconnect.
 
 3. Send node updates with spoofed IP addresses.  The will cause the peer to send Disco (discovery) packets to the spoofed IPs.  But, an insider with access to Coder can easily send packets wherever they want by just directly opening a socket, no need to involve the coordinator.  The coordinator itself never sends disco packets.

File: enterprise/tailnet/pgcoord.go
Comment (spikecurtis) [reply enterprise/tailnet/pgcoord.go:791]: I'm afraid I don't understand what you're suggesting.  Can you elaborate?

**Comment by mtojek** [comment]
> I think our current position is that features supporting multiple Coderd replicas are Enterprise licensed.

I see. It would be awesome to "port" any extra fixes you identified in the existing coordinator to the in-memory one (if you fixed any).

> The goal here was correctness, not increased performance.

Hopefully not decreased :) 

> Send malformed data: This will result in deserialization errors, which triggers the coordinator to disconnect.

Do you think that we should ensure any integrity by signing exchanged data? I agree that this is a topic for an open discussion, but it can be asked during any security review on the customer side.

> as an insider attack

I was thinking also about a different vector. Assuming that the attacker has access to the database, are they able to reconfigure agents or coordinators by spoofing mapping entries directly in the database? I know that the attacker can do other bad things, but was curious if it is possible to perform a truly stealth attack against the customer infra. I apologize if this is a silly question, but it is better to be safe than sorry.

**Comment by spikecurtis** [comment]
> Do you think that we should ensure any integrity by signing exchanged data? I agree that this is a topic for an open discussion, but it can be asked during any security review on the customer side.

These data are exchanged over a websocket, which, for any serious deployment, will be TLS protected, and therefore, integrity protected.

File: enterprise/tailnet/pgcoord.go
Comment (mtojek) [reply enterprise/tailnet/pgcoord.go:795]: Something like this to ensure type-safety on the method declaration level:

```go

func (q *querier) listenClientRaw(_ context.Context, msg []byte, err error) {
   if err != nil && !xerrors.Is(err, pubsub.ErrDroppedMessages) {
     return
   }
   droppedMessagesErr := err.(pubsub.ErrDroppedMessages)

   q.listenClient(msg, droppedMessagesErr)
}

func (q *querier) listenClient(msg []byte, err pubsub.ErrDroppedMessages) {
...
}
```

**Comment by spikecurtis** [comment]
> I was thinking also about a different vector. Assuming that the attacker has access to the database, are they able to reconfigure agents or coordinators by spoofing mapping entries directly in the database? I know that the attacker can do other bad things, but was curious if it is possible to perform a truly stealth attack against the customer infra. I apologize if this is a silly question, but it is better to be safe than sorry.

It's not a silly question, but the answer is that we consider database access as complete compromise of the system.  It would be possible to stealthily attack the system in numerous ways that I won't get into here.

File: enterprise/tailnet/pgcoord.go
Comment (spikecurtis) [reply enterprise/tailnet/pgcoord.go:795]: `listenClient()` is a pubsub listener with error.  The pubsub either sends a message or an error.  If `err` is non-nil, then the `msg` is zero length and not an actual pubsub message.

Right now, the only error that the pubsub can send, and the only one this code handles is `ErrDroppedMessages`.

**Pull request merged**
